src/test/java/bio/overture/ego/token/JwtTest.java

```diff
@@ -79,4 +92,105 @@ public void jwtContainsUserGroups() {
+    assertEquals(appPerms.size(), 3);
+
+    val permScopes =
+        appPerms.stream()
+            .map(AbstractPermissionService::buildScope)
+            .map(Scope::toString)
+            .collect(Collectors.toList());
+    assertEquals(appClient.getClientId(), app.getClientId());
+
+    // assert appClient scopes match created permissions
+    List<String> scopes = new ArrayList<>(appClient.getScope());
+    assertEquals(scopes.size(), permScopes.size());
+    assertTrue(scopes.containsAll(permScopes) && permScopes.containsAll(scopes));
+
+    // getting app with relationships after permissions are set up, create accessToken jwt
+    val appWithRelations = applicationService.getWithRelationships(app.getId());
+    val appToken = tokenService.generateAppToken(appWithRelations);
+    tokenService.isValidToken(appToken);
+    val accessToken = tokenService.getAppAccessToken(appToken);
+    val accessTokenScope = accessToken.getScope();
+
+    // assert jwt scope matches appClient scope
+    assertEquals(scopes.size(), accessTokenScope.size());
+    assertTrue(scopes.containsAll(accessTokenScope));
+    assertTrue(accessTokenScope.containsAll(scopes));
+  }
+
+  @Test
+  @SneakyThrows
+  public void applicationAndGroupPerms_appJwtHasAllResolvedScopes_Success() {
+    val app =
+        entityGenerator.setupApplication("TestCombinedApp", "testsecret", ApplicationType.CLIENT);
+    val group = entityGenerator.setupGroup("Test Group");
+    val policies = entityGenerator.setupPolicies("SONG", "SCORE");
+
+    val permReq1 = PermissionRequest.builder().policyId(policies.get(0).getId()).mask(READ).build();
+    val permReq2 =
+        PermissionRequest.builder().policyId(policies.get(0).getId()).mask(WRITE).build();
+    val permReq3 = PermissionRequest.builder().policyId(policies.get(1).getId()).mask(READ).build();
+
+    // associate app with group
+    groupService.associateApplicationsWithGroup(group.getId(), Arrays.asList(app.getId()));
+
+    // add app and group permissions
+    applicationPermissionService.addPermissions(app.getId(), Arrays.asList(permReq1, permReq3));
+    groupPermissionService.addPermissions(group.getId(), Arrays.asList(permReq2));
+
+    // get permissions, assert count matches
+    val appPerms = applicationService.getByClientId(app.getClientId()).getApplicationPermissions();
+    val groupPerms = groupService.getWithRelationships(group.getId()).getPermissions();
+    assertEquals(appPerms.size(), 2);
+    assertEquals(groupPerms.size(), 1);
+
+    val appClient = applicationService.loadClientByClientId(app.getClientId());
+
+    val permScopes =
+        appPerms.stream()
+            .map(AbstractPermissionService::buildScope)
+            .map(Scope::toString)
+            .collect(Collectors.toList());
+    assertEquals(appClient.getClientId(), app.getClientId());
+
+    // assert appClient scopes match created permissions
+    List<String> scopes = new ArrayList<>(appClient.getScope());
```

Comment:
use `val`. It makes it clear, that the reference is fixed and not changing later on. In some cases, you might do something like the following:
```
Person person = new Person("Joe");

// .... lots of other code //

if (something){
  person = new Person("Mile"); 
}
```

Which cannot be done if `person` is final. Using val makes it clear, that you are not "mutating" the reference later on. It makes it easier to read and review. Otherwise, I search the rest of the code to see how `person` is mutating...to find out its not.